mesonbuild/backend/ninjabackend.py

```diff
@@ -76,48 +132,121 @@ def write(self, outfile):
+        self.rspfile_quote_style = rspfile_quote_style  # rspfile quoting style is 'gcc' or 'cl'
+
+        if self.depfile == '$DEPFILE':
+            self.depfile += '_UNQUOTED'
+
+    @staticmethod
+    def _quoter(x, qf = quote_func):
+        if isinstance(x, NinjaCommandArg):
+            if x.quoting == Quoting.none:
+                return x.s
+            elif x.quoting == Quoting.notNinja:
+                return qf(x.s)
+            elif x.quoting == Quoting.notShell:
+                return ninja_quote(x.s)
+            # fallthrough
+        return ninja_quote(qf(str(x)))
 
     def write(self, outfile):
-        if not self.refcount:
-            return
-
-        outfile.write('rule %s\n' % self.name)
-        if self.rspable:
-            outfile.write(' command = %s @$out.rsp\n' % ' '.join(self.command))
-            outfile.write(' rspfile = $out.rsp\n')
-            outfile.write(' rspfile_content = %s\n' % ' '.join(self.args))
+        if self.rspfile_quote_style == 'cl':
+            rspfile_quote_func = cmd_quote
         else:
-            outfile.write(' command = %s\n' % ' '.join(self.command + self.args))
-        if self.deps:
-            outfile.write(' deps = %s\n' % self.deps)
-        if self.depfile:
-            outfile.write(' depfile = %s\n' % self.depfile)
-        outfile.write(' description = %s\n' % self.description)
-        if self.extra:
-            for l in self.extra.split('\n'):
-                outfile.write(' ')
-                outfile.write(l)
-                outfile.write('\n')
-        outfile.write('\n')
+            rspfile_quote_func = gcc_rsp_quote
+
+        for rsp in ([''] if self.refcount else [] +
+                    ['_RSP'] if self.rsprefcount else []):
+            outfile.write('rule %s%s\n' % (self.name, rsp))
+            if rsp:
+                outfile.write(' command = %s @$out.rsp\n' % ' '.join([self._quoter(x) for x in self.command]))
+                outfile.write(' rspfile = $out.rsp\n')
+                outfile.write(' rspfile_content = %s\n' % ' '.join([self._quoter(x, rspfile_quote_func) for x in self.args]))
+            else:
+                outfile.write(' command = %s\n' % ' '.join([self._quoter(x) for x in (self.command + self.args)]))
+            if self.deps:
+                outfile.write(' deps = %s\n' % self.deps)
+            if self.depfile:
+                outfile.write(' depfile = %s\n' % self.depfile)
+            outfile.write(' description = %s\n' % self.description)
+            if self.extra:
+                for l in self.extra.split('\n'):
+                    outfile.write(' ')
+                    outfile.write(l)
+                    outfile.write('\n')
+            outfile.write('\n')
+
+    def length_estimate(self, infiles, outfiles, elems):
+        # determine variables
+        # this order of actions only approximates ninja's scoping rules, as
+        # documented at: https://ninja-build.org/manual.html#ref_scope
+        vars = {}
```

Comment:
__[Flake8]__ [A001] "vars" is a python builtin and is being shadowed, consider renaming the variable ([link](https://sider.review/gh/repos/19784232/pulls/5245#issue-4043274))

<sub>posted by Sider</sub>